Skip only hop of __torch_function__ at a time

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74727
* #83014
* __->__ #74728

Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

peterbell10 committed:
Skip only hop of __torch_function__ at a time

Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Accept onnx expecttests on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Rebase and move TLS on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Fix possible reference leak on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Fix merge conflicts on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Rebase and fix merge conflicts on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Rebase and fix merge conflicts on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Rebase and fix merge conflicts on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Rebase and fix merge conflicts on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Add deprecations for inplace overloads on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

ezyang commented:
What's the status on this? I'd still like to have it!

peterbell10 committed:
Fix botched merge on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 commented:
This has been blocked on the stack of preceding PRs for a long time and picked up lots of interesting merge conflicts along the way. I need to figure these new test failures out and then it should be good to go (fingers crossed).

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

ezyang commented on torch/csrc/utils/disable_torch_function.cpp:
```diff
@@ -227,6 +272,24 @@ auto check_has_torch_function(PyObject* obj, bool ignore_mode) -> bool {
+  return (
+      !torch::should_skip_torch_function() &&
+      torch::check_has_torch_function(obj));
+}
```
This seems a bit dangerous: this predicate is side-effecting, but it is non-obvious from the name. Easy to accidentally call it twice (thinking that it's idempotent), but actually the next time around it will report true. This may be the source of errors.

peterbell10 replied:
Is the name the main issue, or would you prefer that the caller explicitly checks `should_skip_torch_function`?

ezyang replied:
It's related to https://github.com/pytorch/pytorch/pull/74728#issuecomment-1208446254 . Essentially, it is non-obvious that this function call is what causes the skip boolean to toggle. If it's done via an RAII guard, it's a lot clearer that there is some change (which will get reverted upon exiting the scope, rather than potentially escaping to unbounded scope).

ezyang commented:
@peterbell10, I was thinking over the possible implementations again, and it occurred to me that when we can do https://github.com/pytorch/pytorch/issues/55093#issuecomment-813268742 it seems better to do it this way. My basic fear is that it's easy to accidentally turn off the skipping flag too early.

For BC, we can retain the boolean flag business in Python; but in C++, where we are rewriting the `has_torch_function` calls anyway, we can do the dynamic scoping version of the variable and it seems safer. We might also offer the dynamic scoping version in Python, maybe something like:

```
if g := has_torch_function_scoped(args, kwargs):
  return handle_torch_function(...)
...regular_impl...
```

The g here controls the scope when the skippy flag is set to false; when it goes out of scope the skippy flag goes back to true. 

WDYT?

peterbell10 commented:
```
if g := has_torch_function_scoped(args, kwargs):
  return handle_torch_function(...)
...regular_impl...
```

Can you explain what `g` does exactly here? `g`'s lifetime is only significant in the `handle_torch_function` case which is the non-skipping case anyway, so no flag manipulation isn't required.

peterbell10 commented:
Or is the intention does the lifespan of `g` extend over `...regular_impl...` as well? In that case this wouldn't allow for any `__torch_function__` dispatch within `...regular_impl...`.

ezyang commented:
Walrus operator scope extends beyond the if statement, so it would stay live until you exited from the Python function. When g is in scope, "skip next" is set to False; when it goes out of scope, we reset "skip next" to its value on entry. 

> Or does the lifespan of g extend over ...regular_impl... as well? In that case this wouldn't allow for any __torch_function__ dispatch within ...regular_impl...

Wrong way around. "skip next" is False while g is in scope, so you get to have torch function dispatch in regular impl.

peterbell10 commented:
I think I'm a bit confused because I don't understand what you're trying to fix. I thought you were concerned with "accidentally turn off the skipping flag too early" but if calling `has_torch_function_scope` instantly sets the flag to false then that doesn't change anything wrt setting the flag early.

I'm also not sure why it's desirable to have the flag be true on exit, unless you want it to be a context manager now so we can call multiple functions in a row and have them all skip one level of `__torch_function__`?

ezyang commented:
> I think I'm a bit confused because I don't understand what you're trying to fix. I thought you were concerned with "accidentally turn off the skipping flag too early" but if calling has_torch_function_scope instantly sets the flag to false then that doesn't change anything wrt setting the flag early.

The key is the RAII style guard object that `has_torch_function_scoped` returns. This dictates the scope in which the flag is to be disabled. If you let it go out of scope, then the flag reverts back to the original value (true). In other words, it forces the client to think about *how long* they want the skipping flag to be disabled. If they weren't thinking (and just queried `has_torch_function_scoped` inside a utility function, without returning the guard), the skip flag will revert on exit: no harm done, we will keep skipping until the user explicitly says "ah yes, now I want to call into the underlying function with skips disabled."

> you want it to be a context manager now so we can call multiple functions in a row and have them all skip one level of __torch_function__?

This is certainly true, although it is hard to imagine a situation where you would actually engineer this. No, it's more about bringing back some sort of scoping to the reasoning of the skip flag, the thing I was worried about in https://github.com/pytorch/pytorch/issues/55093#issuecomment-814567929 .

ezyang commented:
In other words, making this dynamic scoping rather than a toggle triggered by a function call is about *scope.* A dynamically scoped handler guarantees that the extent of the behavior change is limited to the dynamic scope of the function call. A boolean flag mutation potentially escapes to an unbounded external scope on return. We have agreed that we are going to add checks to ensure this doesn't escape, but the place where we put the check is precisely where a dynamically scoped guard could have been inserted instead, with the added benefit of being more clear. The only place we can't do this is in legacy Python has torch function tests, which I propose we continue to support in the way we decided on here.

ezyang commented on torch/csrc/utils/disable_torch_function.cpp:
```diff
@@ -143,6 +148,46 @@ PyObject* THPModule_disable_torch_function(PyObject* self, PyObject* a) {
+      PyObject_Call(func, py_args.ptr(), kwargs));
+  bool prev_skip = at::impl::PythonTorchFunctionTLS::exchange_skip_next(false);
+  TORCH_CHECK(
+      !prev_skip || !result,
```
This doesn't look right. If the inner function call failed, it may have failed before we got to the `has_torch_function` test. So it is best to reset the skip next but then propagate the *original* Python error in this case.

peterbell10 replied:
If the call fails, then the `!result` check passes and no exception is raised.

ezyang replied:
oh oops, misread

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

ezyang commented:
@albanD, @bdhirsh and I had a discussion about the corresponding semantics for torch dispatch, and we were thinking that it would be BC-breaking to change torch function to only skip one hop at a time, so let's not change the default and instead provide a separate API for "I'd like to redispatch this and call reentrantly." One application that would be particularly BC broken is a tracer: with one hop you will now see traces for both the upper and lower levels.

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Fix RedispatchTensor serialization on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

albanD commented:
@ezyang I agree that we do want a redispatch function that will skip one level.
If possible, it would be great to have a common API for torch_function and torch_dispatch for sure. Maybe even torch.library?

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Update on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "Skip only hop of __torch_function__ at a time"


Closes #55093

This changes `Tensor.__torch_function__` to skip only one level of
`__torch_function__` calls. This is achieved by adding a new state variable
`skip_next_torch_function` which is reset to false every time
`has_torch_function` is checked, which works both in C++ and Python.

In the issue we discussed making `DisableTorchFunction` public, but with this
you can just write `super().__torch_function__(func, types, args, kwargs)`
to achieve the intended effect whilst being clear that this is
intended for use in `__torch_function__` implementations only.

[ghstack-poisoned]
